src/check/canonicalize.zig

```diff
@@ -174,302 +276,750 @@ fn bringIngestedFileIntoScope(
+
+            const fresh_num_var = self.can_ir.type_store.fresh();
+            const fresh_prec_var = self.can_ir.type_store.fresh();
+
+            const float_expr = CIR.Expr{
+                .float = .{
+                    .num_var = fresh_num_var,
+                    .precision_var = fresh_prec_var,
+                    .literal = literal,
+                    .value = value,
+                    // TODO shouldn't this be a flex_var?
+                    .bound = types.Num.Compact.Frac.Precision.dec,
+                },
+            };
+
+            return self.can_ir.store.addExpr(.{
+                .expr = float_expr,
+                .region = e.region.toBase(),
+            });
+        },
+        .string => |e| {
+            // Get all the string parts
+            const parts_span = e.parts;
+            const parts = self.parse_ir.store.exprSlice(parts_span);
+
+            // Extract segments from the string
+            const segments = self.extractStringSegments(parts) catch |err| exitOnOom(err);
+            defer self.can_ir.env.gpa.free(segments);
+
+            // Check if this is a simple string (no interpolation)
+            if (isSimpleString(segments)) {
+                // Simple string case - concatenate all plaintext segments
+                const final_string = self.concatenatePlaintextSegments(segments) catch |err| exitOnOom(err);
+                defer self.can_ir.env.gpa.free(final_string);
+
+                return self.createStringLiteral(final_string, e.region.toBase());
+            }
+
+            // Handle string interpolation by desugaring to Str.concat calls
+            // TODO: This requires implementing function calls in the IR first
+            //
+            // The desugaring transforms interpolated strings into nested Str.concat calls:
+            // "Hello ${name}, welcome!" becomes:
+            //   Str.concat (Str.concat "Hello " name) ", welcome!"
+            //
+            // For now, we'll collect all the segments and prepare for future desugaring
+            return self.desugarStringInterpolation(segments, e.region.toBase());
+        },
+        .list => |e| {
+            var items = collections.SafeList(CIR.Expr.Idx).initCapacity(self.can_ir.env.gpa, 0);
+            const items_slice = self.parse_ir.store.exprSlice(e.items);
+
+            for (items_slice) |item| {
+                if (self.canonicalize_expr(item)) |canonicalized| {
+                    _ = items.append(self.can_ir.env.gpa, canonicalized);
+                }
+            }
+
+            const fresh_type_var = self.can_ir.type_store.fresh();
+
+            // Mark the start of scratch expressions for the list
+            const scratch_top = self.can_ir.store.scratchExprTop();
+
+            // Add all canonicalized items to scratch
+            var iter = items.iterIndices();
+            while (iter.next()) |idx| {
+                self.can_ir.store.addScratchExpr(items.get(idx).*);
+            }
```

Comment:
Is there a reason to have the allocation of `items` here? Can we just append straight to scratch?

```zig
const scratch_top = self.can_ir.store.scratchExprTop();

const items_slice = self.parse_ir.store.exprSlice(e.items);

for (items_slice) |item| {
  if (self.canonicalize_expr(item)) |canonicalized| {
    self.can_ir.store.addScratchExpr(canonicalized);
  }
}
```